### PR TITLE
New obfuscator  rules

### PR DESCRIPTION
**Pull request by enovella**
I never wrote any YARA rule until now. They need some "massage" but should work without much effort.

### CODE REVIEW

File: apkid/rules/dex/obfuscators.yara [hunk 1]
```diff
@@ -109,3 +109,28 @@ rule bitwise_antiskid : obfuscator
     is_dex and
     any of them
 }
+
+rule arxan : obfuscator
+{
+  meta:
+    description = "Arxan Dalvik"
```
Comment (CalebFenton) [inline apkid/rules/dex/obfuscators.yara]: Could you add a link to a website describing what this is?
Also, please include a reference sample hash.

File: apkid/rules/elf/packers.yara [hunk 2]
```diff
@@ -156,7 +156,7 @@ rule upx_elf_3_03 : packer
     $copyright = "UPX 3.03 Copyright"
 
   condition:
-	upx_unmodified and $copyright
+	  upx_unmodified and $copyright
```
Comment (CalebFenton) [inline apkid/rules/elf/packers.yara]: Accidental white space?

File: apkid/rules/elf/packers.yara [hunk 3]
```diff
@@ -289,3 +289,30 @@ rule upx_unknown_version_unmodified : packer
     upx_unmodified and
     not upx_compressed_apk
 }
+
+rule ollvm_obfuscator : packer embedded
+{
+  meta:
+    description = "Obfuscator-LLVM ELF embedded in APK"
```
Comment (CalebFenton) [inline apkid/rules/elf/packers.yara]: Could we get a URL and hash here also?

**Comment by CalebFenton** [comment]
Thanks for the contribution. Just need a little extra info to review.

File: apkid/rules/elf/packers.yara (hunk 3 above)
Comment (enovella) [reply apkid/rules/elf/packers.yara]: I already added at https://github.com/rednaga/APKiD/issues/5 Is this enough?
The ollvm detection on shared objects has not been tested. We should review it together.

File: apkid/rules/dex/obfuscators.yara (hunk 1 above)
Comment (enovella) [reply apkid/rules/dex/obfuscators.yara]: It seems that @strazzere can provide more insight here. I will find another sample to share with you guys.
Description: https://www.arxan.com/products/application-protection-mobile/

File: apkid/rules/elf/packers.yara [hunk 4]
```diff
@@ -289,3 +289,28 @@ rule upx_unknown_version_unmodified : packer
     upx_unmodified and
     not upx_compressed_apk
 }
+
+rule ollvm_obfuscator : obfuscator
+{
+  meta:
+    description = "Obfuscator-LLVM native"
```
Comment (CalebFenton) [inline apkid/rules/elf/packers.yara]: The 'native' here should be implied by the file name or type mapped to this detection. For that reason, and to make consistent with other names, how about just "Obfuscator-LLVM".

Also, could you add a comment above the description with a URL to the best reference site for this. And can you do the same for Arxan? This way, if someone wants to know "WTF IS ARXAN?" they can look at the rules and find the site.

File: apkid/rules/elf/packers.yara (hunk 2 above)
Comment (CalebFenton) [reply apkid/rules/elf/packers.yara]: Could you fix it?

File: apkid/rules/dex/obfuscators.yara (hunk 1 above)
Comment (CalebFenton) [reply apkid/rules/dex/obfuscators.yara]: Thanks, that's perfect. Could you add this as a comment above description?

File: apkid/rules/elf/packers.yara [hunk 5]
```diff
@@ -289,3 +289,28 @@ rule upx_unknown_version_unmodified : packer
...
+    $based_on = "(based on Obfuscator-LLVM"
+
+  condition:
+    all of them
+}
+
+rule arxan_native : obfuscator
+{
+  meta:
+    description = "Arxan native"
```
Comment (CalebFenton) [inline apkid/rules/elf/packers.yara]: This should use the same name as the DEX version, which should just be the name of the product and _possibly_ a version range, if possible, in general. It's extra information that should be implied by the file type.

File: apkid/rules/elf/packers.yara [hunk 6]
```diff
@@ -289,3 +289,28 @@ rule upx_unknown_version_unmodified : packer
...
+
+  strings:
+    $jni_export = "Java_xxxxxx_"
```
Comment (CalebFenton) [inline apkid/rules/elf/packers.yara]: Is there anything else that could be used here? Perhaps opcodes from a decryption method or something?

File: apkid/rules/elf/packers.yara (hunk 2 above)
Comment (enovella) [reply apkid/rules/elf/packers.yara]: done!

File: apkid/rules/dex/obfuscators.yara (hunk 1 above)
Comment (enovella) [reply apkid/rules/dex/obfuscators.yara]: done!

File: apkid/rules/elf/packers.yara (hunk 4 above)
Comment (enovella) [reply apkid/rules/elf/packers.yara]: - Removed "native" and "dalvik" following the style.
- Added tag "info="http://url.com" for obfuscators.

File: apkid/rules/elf/packers.yara (hunk 6 above)
Comment (enovella) [reply apkid/rules/elf/packers.yara]: I never saw it before. I have absolutely no idea how to fingerprint this.

File: apkid/rules/apk/packers.yara [hunk 7]
```diff
@@ -352,7 +352,7 @@ rule kony : packer
 {
   meta:
     description = "Kony"
-	// http://www.kony.com/
+	  // http://www.kony.com/
```
Comment (CalebFenton) [inline apkid/rules/apk/packers.yara]: Please remove these added white spaces

File: apkid/rules/dex/obfuscators.yara [hunk 8]
```diff
@@ -69,7 +69,7 @@ rule dexguard : obfuscator
 rule dexprotector : obfuscator
 {
   meta:
-    description = "DexProtect"
+    description = "DexProtector"
```
Comment (CalebFenton) [inline apkid/rules/dex/obfuscators.yara]: Please limit the scope of your changes to one "thing" per PR. Please move this change to a separate PR. Also, if you think this name should be changed, please explain why in that PR.

File: apkid/rules/dex/obfuscators.yara [hunk 9]
```diff
@@ -109,3 +109,34 @@ rule bitwise_antiskid : obfuscator
     is_dex and
     any of them
 }
+
+rule arxan : obfuscator
+{
+  meta:
+    description = "Arxan"
+    info = "https://www.arxan.com/products/application-protection-mobile"
```
Comment (CalebFenton) [inline apkid/rules/dex/obfuscators.yara]: I think it's a good to have the URL as part of the meta data, but I think the key should be URL.

File: apkid/rules/apk/packers.yara (hunk 7 above)
Comment (enovella) [reply apkid/rules/apk/packers.yara]: done

File: apkid/rules/dex/packers.yara [hunk 10]
```diff
@@ -70,3 +70,25 @@ rule medusah_appsolid_dex : packer
   condition:
     is_dex and $loader and $main_activity
 }
+
+rule apkguard : packer
```
Comment (CalebFenton) [inline apkid/rules/dex/packers.yara]: I love all the stuff you're submitting, but let's keep everything organized into 1 PR = 1 thing. Makes it easier to review and get stuff merged.

File: apkid/rules/dex/packers.yara [hunk 11]
```diff
@@ -70,3 +70,25 @@ rule medusah_appsolid_dex : packer
...
+{
+  meta:
+    description = "ApkGuard"
+    info = "http://apkguard.io/"
+    // "Base64.decode("UEsDBB..."
```
Comment (CalebFenton) [inline apkid/rules/dex/packers.yara]: What's this comment for?

File: apkid/rules/dex/packers.yara [hunk 12]
```diff
@@ -70,3 +70,25 @@ rule medusah_appsolid_dex : packer
...
+    $getdeclaredmethod = "getDeclaredMethod"
+    $import = "java/lang/reflect/InvocationTargetException"
+    $getfilesdir = "getFilesDir"
+    $dex = "UEsDBB"
+    $b64decode = "Base64"
+
+  condition:
+    is_dex and
+    all of them
+}
```
Comment (CalebFenton) [inline apkid/rules/dex/packers.yara]: Nitpick, but could you always end files with a newline?

File: apkid/rules/elf/obfuscators.yara [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+ * GNU General Public License Usage
+ * --------------------------------
+ * Alternatively, this file may be used under the terms of the GNU General
+ * Public License version 3.0 as published by the Free Software Foundation
+ * and appearing in the file LICENSE.GPL included in the packaging of this
+ * file. Please visit http://www.gnu.org/copyleft/gpl.html and review the
+ * information to ensure the GNU General Public License version 3.0
+ * requirements will be met.
+ *
+ **/
```
Comment (CalebFenton) [inline apkid/rules/elf/obfuscators.yara]: Nice :D

File: apkid/rules/elf/obfuscators.yara [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+
+
+rule ollvm_v3_4 : obfuscator
```
Comment (CalebFenton) [inline apkid/rules/elf/obfuscators.yara]: I really like these. Can you move them to another PR?

File: apkid/rules/elf/obfuscators.yara [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+{
+  meta:
+    description = "Obfuscator-LLVM"
+    info = "https://github.com/obfuscator-llvm/obfuscator/wiki"
+
+  strings:
+    $ollvm = "Obfuscator-LLVM"
+
+  condition:
+    all of them
```
Comment (CalebFenton) [inline apkid/rules/elf/obfuscators.yara]: I think this rule is too broad. If an app just had that as a string, we'd detect it. I think there should be a little bit more, but I agree that this generic, version-less rule should exist.

Also, the condition should say "and not any of the other ollvm rules"

File: apkid/rules/elf/obfuscators.yara [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+// This rule needs to be highly improved.
```
Comment (CalebFenton) [inline apkid/rules/elf/obfuscators.yara]: Then let's not add it in just yet and put it in a separate PR so we can discuss there.

File: apkid/rules/dex/packers.yara (hunk 11 above)
Comment (enovella) [reply apkid/rules/dex/packers.yara]: Removed but it was there to see if we could detect this string in Dalvik code. To be more precise actually.

File: apkid/rules/dex/packers.yara (hunk 12 above)
Comment (enovella) [reply apkid/rules/dex/packers.yara]: Of course from now on. Np

**Comment by enovella** [comment]
Split into several PRs
